File: src/main/java/org/cerion/stocklist/web/api/GoogleFinance.java
package org.cerion.stocklist.web.api;

import org.cerion.stocklist.model.Quote;
import org.cerion.stocklist.model.Symbol;
import org.cerion.stocklist.web.Tools;
import org.cerion.stocklist.web.json.Json;
import org.cerion.stocklist.web.json.JsonArray;
import org.cerion.stocklist.web.json.JsonObject;

import java.util.ArrayList;
import java.util.List;

public class GoogleFinance {

    public Quote getQuote(String symbol) {
        String url = "https://finance.google.com/finance?q=" + symbol + "&output=json";
        Json json = getJson(url);

        // If invalid then result is an array
        if (!(json instanceof JsonArray))
            return null;

        JsonArray arr = (JsonArray)json;
        JsonObject obj = arr.getObject(0);

        Quote quote = new Quote(obj.getString("symbol"));
        quote.setName(obj.getString("name"));
        quote.setExchange(obj.getString("exchange"));

        quote.setLastTrade(obj.getFloat("l"));
        quote.setChange(obj.getFloat("c"));
        quote.setChangePercent(obj.getFloat("cp"));

        quote.setMarketCap(obj.getString("mc"));
        quote.setPeRatio(obj.getFloat("pe"));
        quote.setDividendYield(obj.getFloat("dy"));
        quote.setHigh52(obj.getFloat("hi52"));
        quote.setLow52(obj.getFloat("lo52"));
        quote.setBeta(obj.getFloat("beta"));
        quote.setSector(obj.getString("sname"));
        quote.setEps(obj.getFloat("eps"));

        String volume = obj.getString("vo");
        int mult = 1;
        if (volume.contains("M"))
            mult = 1000000;
        volume = volume.replace("M","");
        Double v = Double.parseDouble(volume) * mult;
        quote.setVolume(v.longValue());

        return quote;
    }

    public List<Symbol> getSymbols(List<String> symbols) {
        String tickers = String.join(",", symbols);
        String url = "https://finance.google.com/finance?q=" + tickers + "&output=json";

        JsonObject json = (JsonObject)getJson(url);
        JsonArray results = json.getArray("searchresults");

        if (results.size() != symbols.size())
            return null;

        List<Symbol> result = new ArrayList<>();
        for(int i = 0; i < results.size(); i++) {
            JsonObject entry = results.getObject(i);

            Symbol s = new Symbol(entry.getString("ticker"), entry.getString("title"), entry.getString("exchange"));
            result.add(s);
        }

        return null;
    }

    private Json getJson(String url) {
        String data = Tools.getURL(url);

        int i = 0;
        while (i < data.length()) {
            char c = data.charAt(i);
            if (c != '\r' && c != '\n' && c != '/')
                break;

            i++;
        }

        return Json.parse(data.substring(i));
    }
}

File: src/main/kotlin/web/api/GoogleFinance.kt
package org.cerion.stocklist.web.api

import org.cerion.stocklist.model.Quote
import org.cerion.stocklist.model.Symbol
import org.cerion.stocklist.web.Tools
import org.cerion.stocklist.web.json.Json
import org.cerion.stocklist.web.json.JsonArray
import org.cerion.stocklist.web.json.JsonObject

import java.util.ArrayList

class GoogleFinance {

    fun getQuote(symbol: String): Quote? {
        val url = "https://finance.google.com/finance?q=$symbol&output=json"
        val json = getJson(url) as? JsonArray ?: return null

        // If invalid then result is an array

        val obj = json.getObject(0)

        val quote = Quote(obj.getString("symbol"))
        quote.name = obj.getString("name")
        quote.exchange = obj.getString("exchange")

        quote.lastTrade = obj.getFloat("l")
        quote.change = obj.getFloat("c")
        quote.changePercent = obj.getFloat("cp")

        quote.marketCap = obj.getString("mc")
        quote.peRatio = obj.getFloat("pe")
        quote.dividendYield = obj.getFloat("dy")
        quote.high52 = obj.getFloat("hi52")
        quote.low52 = obj.getFloat("lo52")
        quote.beta = obj.getFloat("beta")
        quote.sector = obj.getString("sname")
        quote.eps = obj.getFloat("eps")

        var volume = obj.getString("vo")
        var mult = 1
        if (volume.contains("M"))
            mult = 1000000
        volume = volume.replace("M", "")
        val v = java.lang.Double.parseDouble(volume) * mult
        quote.volume = v.toLong()

        return quote
    }

    fun getSymbols(symbols: List<String>): List<Symbol>? {
        val tickers = symbols.joinToString(",")
        val url = "https://finance.google.com/finance?q=$tickers&output=json"

        val json = getJson(url) as JsonObject?
        val results = json!!.getArray("searchresults")

        if (results.size() != symbols.size)
            return null

        val result = ArrayList<Symbol>()
        for (i in 0 until results.size()) {
            val entry = results.getObject(i)

            val s = Symbol(entry.getString("ticker"), entry.getString("title"), entry.getString("exchange"))
            result.add(s)
        }

        return null
    }

    private fun getJson(url: String): Json? {
        val data = Tools.getURL(url)

        var i = 0
        while (i < data!!.length) {
            val c = data[i]
            if (c != '\r' && c != '\n' && c != '/')
                break

            i++
        }

        if (data.isNotEmpty())
            return Json.parse(data.substring(i))

        return null
    }
}
